File: src/main/java/com/jloom/registry/ArchetypeRegistry.java
package com.jloom.registry;

import org.yaml.snakeyaml.LoaderOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.SafeConstructor;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public final class ArchetypeRegistry {

    private static final String INDEX_RESOURCE = "/archetypes/archetypes.yml";

    private final Map<String, ArchetypeManifest> byId;

    private ArchetypeRegistry(Map<String, ArchetypeManifest> byId) {
        this.byId = byId;
    }

    @SuppressWarnings("unchecked")
    public static ArchetypeRegistry loadBundled() {
        Yaml yaml = new Yaml(new SafeConstructor(new LoaderOptions()));
        Map<String, ArchetypeManifest> result = new LinkedHashMap<>();

        try (InputStream indexStream = ArchetypeRegistry.class.getResourceAsStream(INDEX_RESOURCE)) {
            if (indexStream == null) {
                return new ArchetypeRegistry(Map.of());
            }
            Map<String, Object> index = yaml.load(indexStream);
            List<String> ids = (List<String>) index.getOrDefault("archetypes", List.of());

            for (String id : ids) {
                String path = "/archetypes/" + id + ".yml";
                try (InputStream in = ArchetypeRegistry.class.getResourceAsStream(path)) {
                    if (in == null) {
                        throw new IllegalStateException("Listed in archetypes.yml but missing: " + path);
                    }
                    Map<String, Object> raw = yaml.load(in);
                    List<String> modules = (List<String>) raw.getOrDefault("modules", List.of());
                    Map<String, String> answers = new LinkedHashMap<>();
                    Object answersRaw = raw.get("answers");
                    if (answersRaw instanceof Map<?, ?> m) {
                        m.forEach((k, v) -> answers.put(String.valueOf(k), String.valueOf(v)));
                    }
                    result.put(id, new ArchetypeManifest(id, modules, answers));
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }

        return new ArchetypeRegistry(Map.copyOf(result));
    }

    public List<ArchetypeManifest> all() {
        return List.copyOf(byId.values());
    }

    public Optional<ArchetypeManifest> find(String id) {
        return Optional.ofNullable(byId.get(id));
    }
}

File: node/src/registry/archetypeRegistry.ts
import yaml from "js-yaml";
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { catalogRoot } from "./catalogRoots.ts";
import type { ArchetypeManifest } from "./types.ts";

const ARCHETYPES_DIR = path.join(catalogRoot(), "archetypes");
const INDEX_PATH = path.join(ARCHETYPES_DIR, "archetypes.yml");

export class ArchetypeRegistry {
  private readonly byId: Map<string, ArchetypeManifest>;

  private constructor(byId: Map<string, ArchetypeManifest>) {
    this.byId = byId;
  }

  static loadBundled(): ArchetypeRegistry {
    if (!existsSync(INDEX_PATH)) {
      return new ArchetypeRegistry(new Map());
    }

    const index = yaml.load(readFileSync(INDEX_PATH, "utf8")) as { archetypes?: string[] } | undefined;
    const ids = index?.archetypes ?? [];
    const result = new Map<string, ArchetypeManifest>();

    for (const id of ids) {
      const filePath = path.join(ARCHETYPES_DIR, `${id}.yml`);
      if (!existsSync(filePath)) {
        throw new Error(`Listed in archetypes.yml but missing: ${filePath}`);
      }
      const raw = yaml.load(readFileSync(filePath, "utf8")) as
        | { modules?: string[]; answers?: Record<string, unknown> }
        | undefined;
      const modules = raw?.modules ?? [];
      const answers: Record<string, string> = {};
      if (raw?.answers !== undefined && raw.answers !== null && typeof raw.answers === "object") {
        for (const [key, value] of Object.entries(raw.answers)) {
          answers[key] = String(value);
        }
      }
      result.set(id, { id, modules, answers });
    }

    return new ArchetypeRegistry(result);
  }

  all(): ArchetypeManifest[] {
    return [...this.byId.values()];
  }

  find(id: string): ArchetypeManifest | undefined {
    return this.byId.get(id);
  }
}
